### PR TITLE
[SYCL-MLIR] Change the default behavior to generate all SYCL functions if not explicitly mark unsupported

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Removed the `supportedFuncs` list, and replaced with a `unsupportedFuncs` list. 
We mark functions that cannot yet supported in the `unsupportedFuncs` list, and by default generate all other SYCL functions.
When `-gen-all-sycl-funcs` is used, then all SYCL functions are generated.
This PR also includes some small changes to improve codegen.

Signed-off-by: Tsang, Whitney [whitney.tsang@intel.com](mailto:whitney.tsang@intel.com)

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
Depends on https://github.com/intel/llvm/pull/7457 and https://github.com/intel/llvm/pull/7459.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 1]
```diff
@@ -75,124 +75,43 @@ MLIRScanner::MLIRScanner(MLIRASTConsumer &Glob,
...
-
-  // Other SYCL functions:
-  supportedFuncs.insert(
-      "_ZN4sycl3_V13ext6oneapi22accessor_property_listIJEE12has_propertyINS2_"
-      "8property9no_offsetEEEbPNSt9enable_ifIXsr24is_compile_time_propertyIT_"
-      "EE5valueEvE4typeE");
-  supportedFuncs.insert(
-      "_ZNK4sycl3_V18accessorIiLi1ELNS0_6access4modeE1025ELNS2_"
+void MLIRScanner::initUnsupportedFunctions() {
+  // FIXME: -no-mangled-function-name: cgeist:
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:84]: @victor-eds Can you help look into this `-no-mangled-function-name` issue?

**Review by etiotto — APPROVED** [review]
LGTM

File: polygeist/tools/cgeist/Test/Verification/sycl/functions.cpp [hunk 2]
```diff
@@ -218,10 +219,6 @@ SYCL_EXTERNAL void id_get_2(const sycl::id<1> id, int dimension) {
 // CHECK-MLIR:           %arg0: memref<?x!sycl_id_1_> {llvm.align = 8 : i64, llvm.byval = !sycl_id_1_, llvm.noundef}
 // CHECK-MLIR: %{{.*}} = "sycl.id.get"(%{{.*}}) {BaseType = memref<?x!sycl_id_1_, 4>, FunctionName = @"operator unsigned long", MangledFunctionName = @_ZNK4sycl3_V12idILi1EEcvmEv, TypeName = @id} : (memref<?x!sycl_id_1_, 4>) -> i64
 
-// CHECK-LLVM-LABEL: define spir_func void @_Z8id_get_3N4sycl3_V12idILi1EEE(
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/functions.cpp:221]: Some CHECK-LLVM lines are removed, as calls are inlined, so not very useful to check.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 3]
```diff
@@ -274,11 +274,11 @@ class MLIRScanner : public clang::StmtVisitor<MLIRScanner, ValueCategory> {
...
   LowerToInfo &LTInfo;
 
-  // Initialize a whitelist of SYCL functions to emit instead just the
+  // Initialize a exclude list of SYCL functions to emit instead just the
   // declaration. Eventually, this list should be removed.
-  void initSupportedFunctions();
-  bool isSupportedFunctions(std::string Name) const {
-    return supportedFuncs.contains(Name);
+  void initUnsupportedFunctions();
+  bool isUnsupportedFunctions(std::string Name) const {
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:280]: NIT: `isUnsupportedFunction`

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 1 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:84]: I'll look into this. I think we can merge this one and remove this from the list when the problem is solved.

**Review by victor-eds — APPROVED** [review]
LGTM after NIT

**Pull request merged**